backend-services/user-verification-service/src/main/java/org/cardano/foundation/voting/service/discord/DefaultDiscordUserVerificationService.java

```diff
@@ -119,58 +125,98 @@ public Either<Problem, DiscordStartVerificationResponse> startVerification(Strin
+            return Either.left(Problem.builder()
+                    .withTitle("MISSING_WALLET_TYPE")
+                    .withDetail("Wallet type must be specified.")
+                    .withStatus(BAD_REQUEST)
+                    .build());
+        }
 
-            return Either.left(eventDetails.getLeft());
+        Either<Problem, EventSummary> eventValidationResult = validateEvent(eventId);
+        if (eventValidationResult.isLeft()) {
+            return Either.left(eventValidationResult.getLeft());
         }
 
-        var maybeEvent = eventDetails.get();
-        if (maybeEvent.isEmpty()) {
-            log.warn("Event not found:{}", eventId);
+        EventSummary eventSummary = eventValidationResult.get();
+
+        if (!eventSummary.userBased() && walletIdType.get() == WalletType.KERI) {
+            log.warn("Keri wallet not supported for BALANCE or STAKE type event");
 
             return Either.left(Problem.builder()
-                    .withTitle("EVENT_NOT_FOUND")
-                    .withDetail("Event not found, eventId:" + eventId)
+                    .withTitle("WALLET_NOT_SUPPORTED")
+                    .withDetail("Keri wallet not supported for BALANCE or STAKE type event")
                     .withStatus(BAD_REQUEST)
                     .build());
         }
 
-        var event = maybeEvent.orElseThrow();
+        switch (walletIdType.get()) {
+            case CARDANO:
+                return handleCardanoVerification(request, eventId, request.getWalletId());
+            case KERI:
+                return handleKeriVerification(request, eventId, request.getWalletId());
+            default:
+                return Either.left(Problem.builder()
+                        .withTitle("UNSUPPORTED_WALLET_TYPE")
+                        .withDetail("The specified wallet type is not supported.")
+                        .withStatus(BAD_REQUEST)
+                        .build());
+        }
+    }
 
-        if (event.finished()) {
-            log.warn("Event already finished:{}", eventId);
+    private Either<Problem, EventSummary> validateEvent(String eventId) {
+        Either<Problem, Optional<EventSummary>> eventDetailsResult = chainFollowerClient.findEventById(eventId);
+
+        if (eventDetailsResult.isLeft()) {
+            return Either.left(eventDetailsResult.getLeft());
+        }
+
+        return eventDetailsResult.get()
+                .map(Either::<Problem, EventSummary>right)
+                .orElseGet(() -> Either.left(Problem.builder()
+                        .withTitle("EVENT_NOT_FOUND")
+                        .withDetail("Event not found, eventId: " + eventId)
+                        .withStatus(BAD_REQUEST)
+                        .build()));
+    }
 
+    private Either<Problem, IsVerifiedResponse> handleCardanoVerification(DiscordCheckVerificationRequest request, String eventId, String walletId) {
+
+        String signature = request.getCoseSignature().orElse(null);
```

Comment:
I tend to do this like this;
signatureM

and then

if (signatureM.isEmpty())

this is not most monadic way but generally try to avoid nulls in java these days where possible